Added BM25Transformer and BM25Vectorizer to feature_extraction.text. …

BasilBeirouti opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue

<!-- Example: Fixes #1234 -->
#### What does this implement/fix? Explain your changes.
#### Any other comments?

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

…Also added a test to check validity of BM25 calculation.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/scikit-learn/scikit-learn/6973)

<!-- Reviewable:end -->

BasilBeirouti committed:
Merge pull request #1 from scikit-learn/master

merge from original

BasilBeirouti committed:
Added BM25Transformer and BM25Vectorizer to feature_extraction.text. Also added a test to check validity of BM25 calculation.

jnothman commented on sklearn/feature_extraction/text.py:
```diff
@@ -1337,3 +1489,292 @@ def transform(self, raw_documents, copy=True):
+        return self._bm25.smooth_idf
+
+
+def fit(self, raw_documents, y=None):
```
This indentation isn't right.

BasilBeirouti committed:
update text.py

fixed inconsistent indentation starting at line 1708 in text.py all the way tot he end (line 1770)

jnothman commented:
Check the log of TravisCI (under "details"). Just above the coverage listing is a list of failing test cases. Alternatively, run the tests on your local machine.

BasilBeirouti commented:
Hi Joel,

I can see now that these are the same tests that are already included in
the scikit-learn source code, but I'm still not sure how to run them. Is
there a guide or document online that explains this?

Most of the errors that are being thrown don't make any sense to me.

For example, why is the method sign not found here?
======================================================================
ERROR: sklearn.feature_extraction.tests.test_text.test_bm25_no_smoothing
----------------------------------------------------------------------
Traceback (most recent call last):
  File
"/home/travis/build/scikit-learn/scikit-learn/testvenv/local/lib/python2.7/site-packages/nose/case.py",
line 197, in runTest
    self.test(*self.arg)
  File
"/home/travis/sklearn_build_ubuntu/scikit-learn/sklearn/feature_extraction/tests/test_text.py",
line 316, in test_bm25_no_smoothing
    bm25_act = tr.fit_transform(X).toarray()
  File "/home/travis/sklearn_build_ubuntu/scikit-learn/sklearn/base.py",
line 471, in fit_transform
    return self.fit(X, **fit_params).transform(X)
  File
"/home/travis/sklearn_build_ubuntu/scikit-learn/sklearn/feature_extraction/text.py",
line 1045, in transform
    binary = X.sign()
  File "/usr/lib/python2.7/dist-packages/scipy/sparse/base.py", line 384,
in __getattr__
    raise AttributeError(attr + " not found")
AttributeError: sign not found


Why is it complaining that lengths must be the same size as array X?
Lengths is a intermediate variable that I am using to get document lengths,
obviously it should be of size n_samples not n_features x n_samples.

======================================================================
ERROR:
sklearn.tests.test_common.test_non_meta_estimators('BM25Transformer',
<class 'sklearn.feature_extraction.text.BM25Transformer'>)
----------------------------------------------------------------------
Traceback (most recent call last):
  File
"/home/travis/build/scikit-learn/scikit-learn/testvenv/local/lib/python2.7/site-packages/nose/case.py",
line 197, in runTest
    self.test(*self.arg)
  File
"/home/travis/sklearn_build_ubuntu/scikit-learn/sklearn/utils/estimator_checks.py",
line 520, in check_transformer_general
    _check_transformer(name, Transformer, X, y)
  File
"/home/travis/sklearn_build_ubuntu/scikit-learn/sklearn/utils/estimator_checks.py",
line 569, in _check_transformer
    transformer.fit(X, y_)
  File
"/home/travis/sklearn_build_ubuntu/scikit-learn/sklearn/feature_extraction/text.py",
line 999, in fit
    lengths = np.array(X.sum(axis=1)).reshape((n_samples,))
ValueError: total size of new array must be unchanged

I'm a little overwhelmed and not sure how to start fixing this... I also
found some bugs in my pull request outside of anything raised by these
tests which I've already fixed. I would really appreciate some direction.

Sincerely,
Basil Beirouti

maniteja123 commented:
Hi, I am not a core dev but it seems that the test failures are due to ``sign`` function called on sparse matrices which doesn't exist and the other error might be due to adding ``smooth_idf`` for smoothening of the tf-idf term so ``X.shape[0]`` might not be equal to ``n_samples`` now. And probably [this](http://scikit-learn.org/stable/developers/advanced_installation.html) page gives the information regarding running of  test suite. Cheers.

BasilBeirouti commented:
but csr_matrix.sign() does exist! 
http://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.sign.html#scipy.sparse.csr_matrix.sign

thanks for the link to testing information. I'm still unsure of what is causing these errors though

jnothman commented:
the method sign doesn't exist for sparse matrices, nor for numpy arrays
AFAIK. Do you just mean (X >= 0)?

The other problem relates to reshape requiring the overall number of
elements being the same, and apparently (I've not checked why this may be
in your code) X.sum(axis=1) does not (always) have n_samples elements.

On 13 July 2016 at 04:38, Basil Beirouti <notifications@github.com> wrote:

> Hi Joel,
>
> I can see now that these are the same tests that are already included in
> the scikit-learn source code, but I'm still not sure how to run them. Is
> there a guide or document online that explains this?
>
> Most of the errors that are being thrown don't make any sense to me.
>
> For example, why is the method sign not found here?
> ======================================================================
> ERROR: sklearn.feature_extraction.tests.test_text.test_bm25_no_smoothing
> ----------------------------------------------------------------------
> Traceback (most recent call last):
> File
>
> "/home/travis/build/scikit-learn/scikit-learn/testvenv/local/lib/python2.7/site-packages/nose/case.py",
> line 197, in runTest
> self.test(*self.arg)
> File
>
> "/home/travis/sklearn_build_ubuntu/scikit-learn/sklearn/feature_extraction/tests/test_text.py",
> line 316, in test_bm25_no_smoothing
> bm25_act = tr.fit_transform(X).toarray()
> File "/home/travis/sklearn_build_ubuntu/scikit-learn/sklearn/base.py",
> line 471, in fit_transform
> return self.fit(X, **fit_params).transform(X)
> File
>
> "/home/travis/sklearn_build_ubuntu/scikit-learn/sklearn/feature_extraction/text.py",
> line 1045, in transform
> binary = X.sign()
> File "/usr/lib/python2.7/dist-packages/scipy/sparse/base.py", line 384,
> in __getattr__
> raise AttributeError(attr + " not found")
> AttributeError: sign not found
>
>
> Why is it complaining that lengths must be the same size as array X?
> Lengths is a intermediate variable that I am using to get document lengths,
> obviously it should be of size n_samples not n_features x n_samples.
>
> ======================================================================
> ERROR:
> sklearn.tests.test_common.test_non_meta_estimators('BM25Transformer',
> <class 'sklearn.feature_extraction.text.BM25Transformer'>)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
> File
>
> "/home/travis/build/scikit-learn/scikit-learn/testvenv/local/lib/python2.7/site-packages/nose/case.py",
> line 197, in runTest
> self.test(*self.arg)
> File
>
> "/home/travis/sklearn_build_ubuntu/scikit-learn/sklearn/utils/estimator_checks.py",
> line 520, in check_transformer_general
> _check_transformer(name, Transformer, X, y)
> File
>
> "/home/travis/sklearn_build_ubuntu/scikit-learn/sklearn/utils/estimator_checks.py",
> line 569, in _check_transformer
> transformer.fit(X, y_)
> File
>
> "/home/travis/sklearn_build_ubuntu/scikit-learn/sklearn/feature_extraction/text.py",
> line 999, in fit
> lengths = np.array(X.sum(axis=1)).reshape((n_samples,))
> ValueError: total size of new array must be unchanged
>
> I'm a little overwhelmed and not sure how to start fixing this... I also
> found some bugs in my pull request outside of anything raised by these
> tests which I've already fixed. I would really appreciate some direction.
>
> Sincerely,
> Basil Beirouti
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/6973#issuecomment-232139412>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAEz62k0WyBK_BMvO_zVcCoY_0YoOyuYks5qU98agaJpZM4JI7KN>
> .
>

BasilBeirouti committed:
fix indentation and inheritance of bm25vectorizer

fix indentation and inheritance of bm25vectorizer

BasilBeirouti commented:
Thanks for the tips guys.

Joel:

 X.sum(axis=1) would not have n_samples elements, as you said, if X had more than 2 dimensions. I'll have to try adding a shape check in the __init__() method that raises an error if there are more than 2 dimensions. Perhaps that will suffice.

Not exactly X>=0, but for my purposes X>0 would work. However, I thought that the .sign() method would be optimized and therefore faster, however after testing it X>0 is faster, so I will use that no problem. 

Clearly this function sign() either existed or used to exist, here is the documentation for it on numpy and scipy:
http://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.sign.html#scipy.sparse.csr_matrix.sign
http://docs.scipy.org/doc/numpy/reference/generated/numpy.sign.html

These reference pages seem to be generated from the docstrings. I must say I think it's a perfectly reasonable question to ask what's going on when a documented method which I've used before is flagged as not existing by a unit test.

jnothman commented:
I'll admit I've not checked the details, just tried quickly to point you in
the right direction. numpy.sign exists. numpy.ndarray.sign does not.
Perhaps csr_matrix.sign exists too, in some version of scipy. It's also
likely that X > 0 won't be supported in every version of scipy we want to
support, but X_sign = X.copy(); X_sign.data = np.sign(X.data) might suffice.

On 14 July 2016 at 02:51, Basil Beirouti <notifications@github.com> wrote:

> Thanks for the tips guys.
>
> Joel:
>
> X.sum(axis=1) would not have n_samples elements, as you said, if X had
> more than 2 dimensions. I'll have to try adding a shape check in the
> *init*() method that raises an error if there are more than 2 dimensions.
> Perhaps that will suffice.
>
> Not exactly X>=0, but for my purposes X>0 would work. However, I thought
> that the .sign() method would be optimized and therefore faster, however
> after testing it X>0 is faster, so I will use that no problem.
>
> Clearly this function sign() either existed or used to exist, here is the
> documentation for it on numpy and scipy:
>
> http://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.sign.html#scipy.sparse.csr_matrix.sign
> http://docs.scipy.org/doc/numpy/reference/generated/numpy.sign.html
>
> These reference pages seem to be generated from the docstrings. I must say
> I think it's a perfectly reasonable question to ask what's going on when a
> documented method which I've used before is flagged as not existing by a
> unit test.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/6973#issuecomment-232416647>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAEz68KG2PPhLkIWN80-FXFsVl0MYKopks5qVReGgaJpZM4JI7KN>
> .
>

BasilBeirouti committed:
Merge remote-tracking branch 'scikit-learn/master' into mynewfeature

BasilBeirouti committed:
merged new changes from master branch into this branch and resolved merge conflict.

BasilBeirouti committed:
changed X.sign() to X>0

BasilBeirouti commented:
Ah good to know that some versions of scipy don't support X>0 element wise.

I cannot replicate the same nose test results on my local machine as are appearing on the the travis ci build. Instead, I get the output which I attached. 
[nosetests output.txt](https://github.com/scikit-learn/scikit-learn/files/363007/nosetests.output.txt)
bm25 doesn't come up at all. Does anyone know what's going on ? I ran  `nosetests -v sklearn/ ` from the outside directory.

BasilBeirouti committed:
fixed documentation of BM25Transformer

jnothman commented:
Regarding your nose output, it looks like scikit-learn isn't compiled properly. Run `make in` and check all's good there first. Most of those common test failures in Travis, however, are invalid; `CountVectorizer`, `TfidfTransformer` and `TfidfVectorizer` are all listed in `sklearn.utils.testing.DONT_TEST`. (Although personally, I think `TfidfTransformer` should probably be tested there; it just requires binary inputs, but so does `BernoulliNB`.)

Some of them appear to be true failures, nonetheless. And the error in `check_parameters_default_constructible` seems to be a bad limitation of our estimator check there: `unicode` parameters should be allowed..

BasilBeirouti commented:
I have already tried to run `make in`

Note I used `python setup.py develop`  initially to set up my development environment. The way I have it set up now, I have a conda virtual environment which contains the dev version of scikit-learn (which I installed with `python setup.py develop`), and my regular python interpreter (also conda) contains a normal install of the scikit-learn module. Not sure if that helps.

Here's the output of running make in. 
[make in and nosetests output.txt](https://github.com/scikit-learn/scikit-learn/files/365048/make.in.and.nosetests.output.txt)

BasilBeirouti commented:
I activated my conda environment and then ran nose tests, and I think I'm getting the expected output now!

```
[nosetests output after activating environment.txt](https://github.com/scikit-learn/scikit-learn/files/365068/nosetests.output.after.activating.environment.txt)

Basils-MacBook-Pro:scikit-learn basilbeirouti$ source activate mynewenv
(mynewenv) Basils-MacBook-Pro:scikit-learn basilbeirouti$ nosetests -v sklearn/
```

Joel, after I fix the true failing causes, would you suggest I add BM25Vectorizer and BM25Transformer to the DONT_TEST list, and write my own unit tests for them and placing those unit tests in sklearn/feature_extraction/tests/test_text.py instead?

BasilBeirouti committed:
Merge remote-tracking branch 'scikit-learn/master' into mynewfeature

BasilBeirouti committed:
Merge branch 'master' into mynewfeature

BasilBeirouti committed:
added input matrix validation to Bm25Transformer. Added Bm25Transformer and Bm25Vectorizer to DONT_TEST

BasilBeirouti committed:
Merge pull request #2 from scikit-learn/master

merging in upstream changes

BasilBeirouti committed:
Merge pull request #3 from BasilBeirouti/mynewfeature

merging in Bm25 features into master branch

BasilBeirouti committed:
added from __future__ import division to make / operator perform floating point divison unambiguously

BasilBeirouti committed:
Merge pull request #4 from BasilBeirouti/mynewfeature

added from __future__ import division to make / operator perform floa…

BasilBeirouti committed:
added a test for Bm25Transformer smoothing functionality

BasilBeirouti committed:
added a test for Bm25Transformer smoothing functionality

BasilBeirouti committed:
added a test for Bm25Transformer smoothing functionality

BasilBeirouti committed:
added a test for Bm25Transformer smoothing functionality

BasilBeirouti committed:
Merge pull request #5 from BasilBeirouti/mynewfeature

adding assertions for shape of _beta_diag array in Bm25Transformer

BasilBeirouti committed:
added a test for Bm25Transformer smoothing functionality

BasilBeirouti committed:
Merge pull request #6 from BasilBeirouti/mynewfeature

added a test for Bm25Transformer smoothing functionality

BasilBeirouti committed:
Merge pull request #7 from scikit-learn/master

updating my fork

BasilBeirouti committed:
Merge pull request #8 from BasilBeirouti/master

merging updated fork into mynewfeature branch

BasilBeirouti committed:
added a test for Bm25Transformer smoothing functionality

BasilBeirouti committed:
added a test for Bm25Transformer smoothing functionality

amueller commented:
Why did you close?

BasilBeirouti commented:
Hi Andreas,

thanks for checking in! Maybe I'm not using the functionality properly, but
I closed it because the some of the tests failed. Should one keep the pull
request open even if the tests fail ?

One of the Travis tests with Python 2.7 and Scipy 0.9.0 is failing. My
environment is Python 3.5 and the latest version of Scipy so I can't
replicate it. I can't find the exact error I was getting in Travis,
probably because I closed it :( . Conda won't let me install scipy 0.9.0
for some reason, maybe it's too old? I get this:

Error: Package missing in current osx-64 channels


If I can't install the version of scipy that is giving the error, I can't
fix it. So I was basically just guessing at what could have been causing
the problem and submitting pull requests and waiting for the tests to run
to see if my changes fixed it. No success yet. I don't want to keep opening
pull requests and spamming everyone who is getting those notifications so
tonight I'm going to try to figure out how to get scipy 0.9.0 on my local
machine.


Sincerely,

Basil Beirouti

On Fri, Jul 29, 2016 at 2:29 PM, Andreas Mueller <notifications@github.com>
wrote:

> Why did you close?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/6973#issuecomment-236272368>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJ0MI__yZ1IWHlZJSveAgdSAwFFoMO_Hks5qalSqgaJpZM4JI7KN>
> .
>

BasilBeirouti commented:
How does one replicate the same environments that are used on Travis tests?

jnothman commented:
> Should one keep the pull request open even if the tests fail ?

Absolutely. Just make new commits and push to the PR's branch on your fork.

> One of the Travis tests with Python 2.7 and Scipy 0.9.0 is failing

https://travis-ci.org/scikit-learn/scikit-learn/builds/148614520 shows that tests are failing in *all* environments. The errors look like you should be able to test and resolve these issues on your own machine; they are not really dependent on platform.

> How does one replicate the same environments that are used on Travis tests?

Replicating the travis environments locally can be a little tricky, but usually with some expert guidance, the error messages are sufficient.

BasilBeirouti commented:
Hi Joel,

Thanks for the quick response. The errors in 3.5 don't seem to be related to my code additions. I'm concerned about the ValueErrors in the python 2.6 and python 2.7 environments. Something has changed about the way sparse matrices work between scipy versions. 

`ValueError: operands could not be broadcast together with shapes (1,3) (7) `

jnothman commented:
I've not looked in detail, but you may need to do `self._beta_diag.dot(binary).tocsr()` or something...

jnothman commented on sklearn/feature_extraction/text.py:
```diff
@@ -913,6 +916,170 @@ def _make_int_array():
+        binary = X>0
+
+        weightedtfs.data = (self.k + 1) / \
+                           (self.k * self._beta_diag.dot(binary).data / X.data + 1)
```
Unfortunately, I don't think there's any `scipy.sparse` API assurance that `self._beta_diag.dot(binary).indices` is identical to `X.indices`, hence their `data` may not be aligned.

BasilBeirouti replied:
self._beta_diag is created from sp.spdiags() method. I see what you mean about the API not guaranteeing the type of diagonal matrix being created.

Looking at the API docs for that method:
http://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.spdiags.html

Note it says:

> > format : str, optional
> Format of the result. By default (format=None) an appropriate sparse matrix format is returned. This choice is subject to change.

I suspect that a different sparse type is being chosen as a return value in Python 2.7 and 2.6 vs Python 3.5. I added some print statements to check as a quick and dirty way to work through it. But I'm also wondering if the X>0 is breaking something as well. 

I might need to add format="csr"

BasilBeirouti committed:
update text.py with prints  in Bm25Transformer

since I cannot get the same scipy environment locally I need to see what's going on with the array shapes

BasilBeirouti committed:
text.py update Bm25Transformer

binary is behaving unexpectedly

BasilBeirouti committed:
fixing Bm25Transformer

removed prints, forced self._beta_diags to be sparse csr_matrix. Changed implementation of X>0

BasilBeirouti commented:
Alright well I guess it's fixed now. 2.7 and 2.6 pass, 3.5 fails in Travis due to errors that don't appear to be directly related to any of my code or tests. Lots of import errors on the failed build. Can someone shed some light?

jnothman commented:
The Python 3.5 problem is something I've seen on one of my PRs. I've cleared the CI cache and restarted the job.

BasilBeirouti commented:
Thank you! Looks like it's all good.

BasilBeirouti committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into mynewfeature

BasilBeirouti committed:
removed temp variable

BasilBeirouti committed:
update text.py

removed temporary variable

BasilBeirouti committed:
Merge branch 'mynewfeature' of https://github.com/BasilBeirouti/scikit-learn into mynewfeature

BasilBeirouti commented:
Could someone review this and give me some feedback? When can I expect these changes to be merged?

Note that the fit method creates the betas and IDF vectors, and then the transform vector checks to see that those vectors have been created uses them to calculate the Bm25 transform. 

There are many different possible variations on the formula for IDF, I've only implemented one variation. I don't like restricting the user to just one implementation, but I think it would be feature creep to build all of them into the Bm25Transformer and Bm25Vectorizer methods. Since IDF weights are calculated from document frequency, I was thinking we could expose the document frequency vector (df) as another property with a getter, and then create a setter for the IDF vector which allows the user to overwrite the existing IDF weights with their own weighting scheme (calculated from the exposed df values). I think that is a simple and more elegant way than attempting to implement the exhaustive list of possible IDF weights. What do people think?

jnothman commented:
> Could someone review this and give me some feedback? When can I expect these changes to be merged?

Thanks for pinging. To be sure, there's no assurance that any PR will be merged. The review process, which depends on volunteers but aims for high quality, can take a while. We require multiple core devs to agree that something should be merged, both in terms of quality and in terms of value in the core scikit-learn project. Alternatively, if you're anxious to make the tool public, you're very welcome to make this transformer available as an external project using our contrib toolkit. See https://github.com/scikit-learn-contrib/

More specifically, there was, from the outset, some apprehension and uncertainly about BM25's relevance to machine learning. I don't think you can expect this to be merged without a motivating example, either a new entry in the `examples` directory, or an extension to an existing one that illustrates that BM25 can exceed the performance of other feature weighting schemes in a machine learning context.

You also aren't going to get this merged without checking for PEP8 compliance first. At a glance I can see numerous violations. Similarly, I can see numerous incorrect uses of the documentation format: try compiling it. Devs are much more likely to be enthusiastic about reviewing your work if you've taken the effort to read and comply with our contributing guidelines, so that we don't need to spend our time nitpicking over each minor violation.

BasilBeirouti committed:
Merge pull request #9 from scikit-learn/master

update branch

BasilBeirouti committed:
Merge pull request #10 from scikit-learn/master

update branch

BasilBeirouti commented:
Joel that all makes sense and thank you for taking the time to explain. I'm still learning how all of this works.

BasilBeirouti committed:
Merge remote-tracking branch 'scikit-learn/master' into mynewfeature

BasilBeirouti committed:
Merge remote-tracking branch 'origin/mynewfeature' into mynewfeature

BasilBeirouti committed:
pep8 formatting fix: made all lines 79 char or less

ifp-markreinke commented:
Hello Basil,
we have tried out your Bm25Transformer and have the following problem: If we fit() the transformer to our train set and then transform() the test set, we get a dimension mismatch.
Here is an excerpt of the stack trace:

`in transform(self, X, copy)
    140         weightedtfs.data = (self.k + 1) / \
    141                            (self.k * self._beta_diag.dot(
--> 142                                binary).data / X.data + 1)
    143 
    144         bm25 = weightedtfs.dot(self._idf_diag)

ValueError: dimension mismatch
`

BasilBeirouti commented:
Hi Mark,

Thanks for trying it out! can you send me an excerpt of the data and code that reproduces the issue either through email basilbeirouti@gmail.com or through a gist and I'll take a look immediately.

BasilBeirouti committed:
pep8 formatting fix: made all lines 79 char or less

BasilBeirouti committed:
added 2 tests for Bm25Transformer properties

BasilBeirouti committed:
added some asserts and cleaned up code a bit

BasilBeirouti committed:
updated from scikit-learn/master

BasilBeirouti committed:
not sure what's going on now

BasilBeirouti committed:
I think this is working

BasilBeirouti committed:
Merge remote-tracking branch 'scikit-learn/master' into mynewfeature

blakeapm commented:
Would love to have this merged. I am using it all the time, great for info retrieval stuff.

jnothman commented:
@blakeapm, we're not an information retrieval library, and not trying to cater to that use case. Do you have successful uses of this transformation in machine learning?

rth commented:
> we're not an information retrieval library, and not trying to cater to that use case. Do you have successful uses of this transformation in machine learning?

Closing, as it is quite unlikely to get merged being specific to information retriveal. field. Adding `BM25Transformer` to [scikit-learn-extra](https://github.com/scikit-learn-contrib/scikit-learn-extra) or some more IR specific package such as https://github.com/zveryansky/textvec could be usefil though.
